File: scenarios/ambiguous_run.py
"""AMBIGUOUS scenario: a raw, underspecified stakeholder ask with no
acceptance criteria. The requirements agent must (1) detect the ambiguity,
(2) generate multiple concrete candidate interpretations, (3) score and
select one with a documented rationale, and (4) record the rejected
alternatives -- all before the design/implementation stages ever run. This
is the contrast case for `greenfield_run.py`, whose request is already
concrete.

The chosen interpretation is implemented as real input-validation hardening
on POST /api/urls (and, for consistency, its bulk sibling): reject URLs that
target a blocklisted domain, and reserve system-critical short codes so they
can't be squatted.

Run with:  python -m scenarios.ambiguous_run
"""
import os

from scenarios.common import PROJECT_ROOT, new_run_id, print_summary, run_dir, write_report

from orchestrator import policy
from orchestrator.agents import design_agent, docs_agent, implementation_agent, release_agent, requirements_agent, testing_agent
from orchestrator.approvals import AutoApprover
from orchestrator.context import SharedContext
from orchestrator.executor import Executor
from orchestrator.graph import Graph, Node
from orchestrator.observability import EventLog

RAW_REQUEST = "Can we make the short links a bit safer? A couple of users raised concerns."


def _build_shortener_content(project_root: str) -> str:
    path = os.path.join(project_root, "app", "shortener.py")
    with open(path, encoding="utf-8") as f:
        content = f.read()
    if "UnsafeTargetError" in content:
        return content  # already applied (idempotent re-run)

    content = content.replace(
        "from datetime import datetime, timedelta, timezone\n\nfrom app import config, db",
        "from datetime import datetime, timedelta, timezone\nfrom urllib.parse import urlparse\n\nfrom app import config, db",
    )

    content = content.replace(
        "class AliasConflictError(Exception):\n    pass\n",
        "class AliasConflictError(Exception):\n    pass\n\n\n"
        "class UnsafeTargetError(Exception):\n    pass\n\n\n"
        "class ReservedAliasError(Exception):\n    pass\n\n\n"
        'BLOCKED_DOMAINS = {"malicious-example.test", "phishing-example.test"}\n'
        'RESERVED_ALIASES = {"api", "health", "admin", "www"}\n',
    )

    content = content.replace(
        "def create_link(url: str, custom_alias: str | None = None, expires_in_days: int | None = None) -> dict:\n"
        "    if custom_alias:",
        "def create_link(url: str, custom_alias: str | None = None, expires_in_days: int | None = None) -> dict:\n"
        "    domain = urlparse(url).netloc.lower()\n"
        "    if domain in BLOCKED_DOMAINS:\n"
        "        raise UnsafeTargetError(f\"target domain '{domain}' is not allowed\")\n"
        "    if custom_alias and custom_alias.lower() in RESERVED_ALIASES:\n"
        "        raise ReservedAliasError(f\"alias '{custom_alias}' is reserved for internal use\")\n"
        "    if custom_alias:",
    )
    return content


def _build_main_content(project_root: str) -> str:
    path = os.path.join(project_root, "app", "main.py")
    with open(path, encoding="utf-8") as f:
        content = f.read()
    if "UnsafeTargetError" not in content:
        content = content.replace(
            "    try:\n"
            "        link = shortener.create_link(req.url, req.custom_alias, req.expires_in_days)\n"
            "    except shortener.AliasConflictError as e:\n"
            "        raise HTTPException(status_code=409, detail=str(e))\n"
            "    return _to_response(link)",
            "    try:\n"
            "        link = shortener.create_link(req.url, req.custom_alias, req.expires_in_days)\n"
            "    except shortener.AliasConflictError as e:\n"
            "        raise HTTPException(status_code=409, detail=str(e))\n"
            "    except (shortener.UnsafeTargetError, shortener.ReservedAliasError) as e:\n"
            "        raise HTTPException(status_code=422, detail=str(e))\n"
            "    return _to_response(link)",
        )
    if "/api/urls/bulk" in content and "except shortener.AliasConflictError as e:\n            results.append" in content:
        content = content.replace(
            "        except shortener.AliasConflictError as e:\n"
            "            results.append(BulkResultItem(success=False, error=str(e), url=item.url))",
            "        except (shortener.AliasConflictError, shortener.UnsafeTargetError, shortener.ReservedAliasError) as e:\n"
            "            results.append(BulkResultItem(success=False, error=str(e), url=item.url))",
        )
    return content


def _build_test_content() -> str:
    return '''"""Tests for the URL-safety hardening added by the ambiguous-requirement
orchestration scenario (scenarios/ambiguous_run.py)."""


def test_blocked_domain_is_rejected(client):
    resp = client.post("/api/urls", json={"url": "https://malicious-example.test/phish"})
    assert resp.status_code == 422


def test_reserved_alias_is_rejected(client):
    resp = client.post("/api/urls", json={"url": "https://example.com", "custom_alias": "api"})
    assert resp.status_code == 422


def test_safe_url_and_alias_still_work(client):
    resp = client.post("/api/urls", json={"url": "https://example.com/fine", "custom_alias": "my-brand-2"})
    assert resp.status_code == 201


def test_bulk_reports_unsafe_item_as_a_per_item_error_not_a_hard_failure(client):
    resp = client.post(
        "/api/urls/bulk",
        json={"urls": [{"url": "https://malicious-example.test/x"}, {"url": "https://example.com/y"}]},
    )
    assert resp.status_code == 201
    body = resp.json()
    assert body["results"][0]["success"] is False
    assert body["results"][1]["success"] is True
'''


def build_graph(project_root: str) -> Graph:
    g = Graph()

    g.add(Node(
        id="requirements", name="Detect ambiguity and normalize the safety request", stage="requirements",
        depends_on=[], run=lambda ctx, node: requirements_agent.normalize(ctx, node),
    ))

    g.add(Node(
        id="design", name="Design the input-validation hardening", stage="design",
        depends_on=["requirements"],
        run=lambda ctx, node: design_agent.propose(
            ctx, node,
            feature_name=ctx.get_output("requirements")["chosen_interpretation"],
            api_contract=[
                {"method": "POST", "path": "/api/urls",
                 "description": "Rejects blocklisted target domains and reserved short-code aliases with 422."},
                {"method": "POST", "path": "/api/urls/bulk",
                 "description": "Applies the same safety checks per item, consistent with the single-create endpoint."},
            ],
            data_model_changes=["No schema/table change; validation happens in app/shortener.py:create_link."],
            risks=[
                "The domain blocklist is static and requires manual maintenance -- a live threat-intel feed was "
                "considered and explicitly rejected for this change (see requirements decision lineage) as "
                "disproportionate effort for the ask as stated.",
                "Reserved-alias list is hardcoded; adding a new reserved word later requires a code change, "
                "not just configuration.",
            ],
            project_root=project_root,
            mode="modify",
        ),
    ))

    g.add(Node(
        id="implementation", name="Implement validation + tests", stage="implementation",
        depends_on=["design"],
        run=lambda ctx, node: implementation_agent.write_files(
            ctx, node,
            files={
                "app/shortener.py": _build_shortener_content(project_root),
                "app/main.py": _build_main_content(project_root),
                "tests/test_url_safety.py": _build_test_content(),
            },
            project_root=project_root,
        ),
        exit_policies=[policy.no_hardcoded_secrets, policy.dependency_allowlist],
        compensate=implementation_agent.revert_files,
    ))

    g.add(Node(
        id="testing", name="Run the full pytest suite", stage="testing",
        depends_on=["implementation"],
        run=lambda ctx, node: testing_agent.run_tests(ctx, node, test_path="tests/", project_root=project_root),
        exit_policies=[policy.tests_must_pass],
    ))

    g.add(Node(
        id="docs", name="Write API changelog entry", stage="docs",
        depends_on=["implementation"],
        run=lambda ctx, node: docs_agent.write_change_doc(
            ctx, node, design_node_id="design", project_root=project_root, doc_rel_path="docs/API_CHANGELOG.md"
        ),
    ))

    g.add(Node(
        id="release", name="Prepare release (human-approved -- safety-relevant change)", stage="release",
        depends_on=["testing", "docs"], requires_approval=True, parallel_ok=False,
        entry_policies=[policy.make_release_gate(g)],
        run=lambda ctx, node: release_agent.prepare_release(
            ctx, node, project_root=project_root, release_notes_rel_path=f"runs/{ctx.run_id}/RELEASE_NOTES.md"
        ),
    ))

    return g


def main() -> None:
    run_id = new_run_id("ambiguous")
    out_dir = run_dir(run_id)
    ctx = SharedContext(run_id=run_id, scenario="ambiguous: 'make links safer'", raw_request=RAW_REQUEST)
    graph = build_graph(PROJECT_ROOT)
    event_log = EventLog(run_id=run_id, out_dir=out_dir)
    executor = Executor(graph, ctx, event_log, approver=AutoApprover(graph=graph), rollback_on_failure=True)

    result = executor.run()
    event_log.close()
    write_report(result, os.path.join(out_dir, "REPORT.md"))
    print_summary(result)

    req_output = ctx.get_output("requirements")
    print("\n--- requirements agent output ---")
    print(f"  ambiguity_detected: {req_output['ambiguity_detected']}")
    print(f"  chosen_interpretation: {req_output['chosen_interpretation']}")
    print(f"  rejected_interpretations: {req_output['rejected_interpretations']}")
    print(f"\nFull report: runs/{run_id}/REPORT.md")


if __name__ == "__main__":
    main()

File: java/src/main/java/com/example/urlshortener/scenarios/AmbiguousRun.java
package com.example.urlshortener.scenarios;

import com.example.urlshortener.orchestrator.*;
import com.example.urlshortener.orchestrator.agents.*;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * AMBIGUOUS scenario: a raw, underspecified stakeholder ask with no
 * acceptance criteria. Direct port of scenarios/ambiguous_run.py -- see
 * that file's docstring for the full narrative (ambiguity detection,
 * scored candidate interpretations, then real input-validation hardening).
 *
 * Run with: mvn exec:java -Dexec.mainClass=com.example.urlshortener.scenarios.AmbiguousRun
 */
public class AmbiguousRun {

    static final String RAW_REQUEST = "Can we make the short links a bit safer? A couple of users raised concerns.";

    static final String UNSAFE_TARGET_EXCEPTION = """
            package com.example.urlshortener.app.exceptions;

            public class UnsafeTargetException extends RuntimeException {
                public UnsafeTargetException(String message) {
                    super(message);
                }
            }
            """;

    static final String RESERVED_ALIAS_EXCEPTION = """
            package com.example.urlshortener.app.exceptions;

            public class ReservedAliasException extends RuntimeException {
                public ReservedAliasException(String message) {
                    super(message);
                }
            }
            """;

    static final String SERVICE_ANCHOR = """
                public Link createLink(String url, String customAlias, Integer expiresInDays) {
                    String code;
                    boolean isCustom;
            """;

    static final String SERVICE_REPLACEMENT = """
                private static final java.util.Set<String> BLOCKED_DOMAINS =
                        java.util.Set.of("malicious-example.test", "phishing-example.test");
                private static final java.util.Set<String> RESERVED_ALIASES =
                        java.util.Set.of("api", "health", "admin", "www");

                public Link createLink(String url, String customAlias, Integer expiresInDays) {
                    String domain = java.net.URI.create(url).getHost();
                    if (domain != null && BLOCKED_DOMAINS.contains(domain.toLowerCase())) {
                        throw new com.example.urlshortener.app.exceptions.UnsafeTargetException(
                                "target domain '" + domain + "' is not allowed");
                    }
                    if (customAlias != null && RESERVED_ALIASES.contains(customAlias.toLowerCase())) {
                        throw new com.example.urlshortener.app.exceptions.ReservedAliasException(
                                "alias '" + customAlias + "' is reserved for internal use");
                    }
                    String code;
                    boolean isCustom;
            """;

    static String buildServiceContent(Path projectRoot) {
        Path path = projectRoot.resolve("src/main/java/com/example/urlshortener/app/ShortenerService.java");
        String content;
        try {
            content = Files.readString(path);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
        if (content.contains("BLOCKED_DOMAINS")) return content;
        return content.replace(SERVICE_ANCHOR, SERVICE_REPLACEMENT);
    }

    static final String HANDLER_ANCHOR = """
                @ExceptionHandler(IllegalArgumentException.class)
                public ResponseEntity<Map<String, String>> handleBadRequest(IllegalArgumentException e) {
                    return ResponseEntity.status(422).body(Map.of("detail", e.getMessage()));
                }
            }
            """;

    static final String HANDLER_REPLACEMENT = """
                @ExceptionHandler(IllegalArgumentException.class)
                public ResponseEntity<Map<String, String>> handleBadRequest(IllegalArgumentException e) {
                    return ResponseEntity.status(422).body(Map.of("detail", e.getMessage()));
                }

                @ExceptionHandler({
                        com.example.urlshortener.app.exceptions.UnsafeTargetException.class,
                        com.example.urlshortener.app.exceptions.ReservedAliasException.class
                })
                public ResponseEntity<Map<String, String>> handleUnsafe(RuntimeException e) {
                    return ResponseEntity.status(422).body(Map.of("detail", e.getMessage()));
                }
            }
            """;

    static String buildHandlerContent(Path projectRoot) {
        Path path = projectRoot.resolve("src/main/java/com/example/urlshortener/app/GlobalExceptionHandler.java");
        String content;
        try {
            content = Files.readString(path);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
        if (content.contains("UnsafeTargetException")) return content;
        return content.replace(HANDLER_ANCHOR, HANDLER_REPLACEMENT);
    }

    static final String BULK_CATCH_ANCHOR = """
                        } catch (com.example.urlshortener.app.exceptions.AliasConflictException e) {
                            results.add(new BulkResultItem(false, null, e.getMessage(), item.url()));
                        }
            """;

    static final String BULK_CATCH_REPLACEMENT = """
                        } catch (com.example.urlshortener.app.exceptions.AliasConflictException
                                | com.example.urlshortener.app.exceptions.UnsafeTargetException
                                | com.example.urlshortener.app.exceptions.ReservedAliasException e) {
                            results.add(new BulkResultItem(false, null, e.getMessage(), item.url()));
                        }
            """;

    static String buildControllerContent(Path projectRoot) {
        Path path = projectRoot.resolve("src/main/java/com/example/urlshortener/app/UrlController.java");
        String content;
        try {
            content = Files.readString(path);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
        if (!content.contains("/api/urls/bulk") || content.contains("UnsafeTargetException")) return content;
        return content.replace(BULK_CATCH_ANCHOR, BULK_CATCH_REPLACEMENT);
    }

    static String buildSafetyTest() {
        return """
                package com.example.urlshortener.app;

                import com.fasterxml.jackson.databind.JsonNode;
                import com.fasterxml.jackson.databind.ObjectMapper;
                import org.junit.jupiter.api.BeforeEach;
                import org.junit.jupiter.api.Test;
                import org.springframework.beans.factory.annotation.Autowired;
                import org.springframework.boot.test.context.SpringBootTest;
                import org.springframework.boot.test.web.client.TestRestTemplate;
                import org.springframework.boot.test.web.server.LocalServerPort;
                import org.springframework.http.ResponseEntity;
                import org.springframework.jdbc.core.JdbcTemplate;
                import org.springframework.test.context.ActiveProfiles;

                import static org.assertj.core.api.Assertions.assertThat;

                @SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
                @ActiveProfiles("test")
                class SafetyValidationTest {

                    @LocalServerPort
                    private int port;

                    @Autowired
                    private TestRestTemplate rest;

                    @Autowired
                    private JdbcTemplate jdbc;

                    @Autowired
                    private ShortenerService shortener;

                    private final ObjectMapper mapper = new ObjectMapper();

                    @BeforeEach
                    void cleanDb() {
                        jdbc.execute("DELETE FROM clicks");
                        jdbc.execute("DELETE FROM links");
                        shortener.redirectCache.clear();
                    }

                    private String url(String path) {
                        return "http://localhost:" + port + path;
                    }

                    @Test
                    void blockedDomainIsRejected() {
                        ResponseEntity<String> resp = rest.postForEntity(
                                url("/api/urls"),
                                new com.example.urlshortener.app.dto.CreateLinkRequest("https://malicious-example.test/phish", null, null),
                                String.class);
                        assertThat(resp.getStatusCode().value()).isEqualTo(422);
                    }

                    @Test
                    void reservedAliasIsRejected() {
                        ResponseEntity<String> resp = rest.postForEntity(
                                url("/api/urls"),
                                new com.example.urlshortener.app.dto.CreateLinkRequest("https://example.com", "api", null),
                                String.class);
                        assertThat(resp.getStatusCode().value()).isEqualTo(422);
                    }

                    @Test
                    void safeUrlAndAliasStillWork() {
                        ResponseEntity<String> resp = rest.postForEntity(
                                url("/api/urls"),
                                new com.example.urlshortener.app.dto.CreateLinkRequest("https://example.com/fine", "my-brand-2", null),
                                String.class);
                        assertThat(resp.getStatusCode().value()).isEqualTo(201);
                    }

                    @Test
                    void bulkReportsUnsafeItemAsPerItemErrorNotHardFailure() throws Exception {
                        var req = new com.example.urlshortener.app.dto.BulkCreateRequest(java.util.List.of(
                                new com.example.urlshortener.app.dto.BulkCreateItem("https://malicious-example.test/x", null, null),
                                new com.example.urlshortener.app.dto.BulkCreateItem("https://example.com/y", null, null)
                        ));
                        ResponseEntity<String> resp = rest.postForEntity(url("/api/urls/bulk"), req, String.class);
                        assertThat(resp.getStatusCode().value()).isEqualTo(201);
                        JsonNode body = mapper.readTree(resp.getBody());
                        assertThat(body.get("results").get(0).get("success").asBoolean()).isFalse();
                        assertThat(body.get("results").get(1).get("success").asBoolean()).isTrue();
                    }
                }
                """;
    }

    static Graph buildGraph(Path projectRoot) {
        Graph g = new Graph();

        g.add(Node.builder("requirements", "requirements", RequirementsAgent::normalize)
                .name("Detect ambiguity and normalize the safety request").build());

        g.add(Node.builder("design", "design", (ctx, node) -> {
                    @SuppressWarnings("unchecked")
                    String featureName = (String) ctx.getOutput("requirements").get("chosen_interpretation");
                    return DesignAgent.propose(
                            ctx, node, featureName,
                            List.of(
                                    new DesignAgent.Endpoint("POST", "/api/urls",
                                            "Rejects blocklisted target domains and reserved short-code aliases with 422."),
                                    new DesignAgent.Endpoint("POST", "/api/urls/bulk",
                                            "Applies the same safety checks per item, consistent with the single-create endpoint.")
                            ),
                            List.of("No schema/table change; validation happens in ShortenerService.createLink."),
                            List.of(
                                    "The domain blocklist is static and requires manual maintenance -- a live "
                                            + "threat-intel feed was considered and explicitly rejected for this change "
                                            + "(see requirements decision lineage) as disproportionate effort for the ask as stated.",
                                    "Reserved-alias list is hardcoded; adding a new reserved word later requires a "
                                            + "code change, not just configuration."
                            ),
                            projectRoot, "modify"
                    );
                })
                .name("Design the input-validation hardening").dependsOn("requirements").build());

        g.add(Node.builder("implementation", "implementation", (ctx, node) -> {
                    Map<String, String> files = new LinkedHashMap<>();
                    files.put("src/main/java/com/example/urlshortener/app/exceptions/UnsafeTargetException.java", UNSAFE_TARGET_EXCEPTION);
                    files.put("src/main/java/com/example/urlshortener/app/exceptions/ReservedAliasException.java", RESERVED_ALIAS_EXCEPTION);
                    files.put("src/main/java/com/example/urlshortener/app/ShortenerService.java", buildServiceContent(projectRoot));
                    files.put("src/main/java/com/example/urlshortener/app/GlobalExceptionHandler.java", buildHandlerContent(projectRoot));
                    files.put("src/main/java/com/example/urlshortener/app/UrlController.java", buildControllerContent(projectRoot));
                    files.put("src/test/java/com/example/urlshortener/app/SafetyValidationTest.java", buildSafetyTest());
                    return ImplementationAgent.writeFiles(ctx, node, files, projectRoot, List.of());
                })
                .name("Implement validation + tests").dependsOn("design")
                .exitPolicy("no_hardcoded_secrets", Policies::noHardcodedSecrets)
                .exitPolicy("dependency_allowlist", Policies::dependencyAllowlist)
                .compensate(ImplementationAgent::revertFiles)
                .build());

        g.add(Node.builder("testing", "testing", (ctx, node) -> TestingAgent.runTests(ctx, node, projectRoot))
                .name("Run mvn test (compile + test)").dependsOn("implementation")
                .exitPolicy("tests_must_pass", Policies::testsMustPass)
                .build());

        g.add(Node.builder("docs", "docs", (ctx, node) -> DocsAgent.writeChangeDoc(
                        ctx, node, "design", projectRoot, "docs/API_CHANGELOG.md"))
                .name("Write API changelog entry").dependsOn("implementation").build());

        g.add(Node.builder("release", "release", (ctx, node) -> ReleaseAgent.prepareRelease(
                        ctx, node, projectRoot, "runs/" + ctx.runId + "/RELEASE_NOTES.md"))
                .name("Prepare release (human-approved -- safety-relevant change)").dependsOn("testing", "docs")
                .requiresApproval(true).parallelOk(false)
                .entryPolicy("release_requires_clean_policy_history", (ctx, node, r) -> Policies.releaseGate(g).check(ctx, node, r))
                .build());

        return g;
    }

    public static void main(String[] args) {
        String runId = Common.newRunId("ambiguous-java");
        Path outDir = Common.runDir(runId);
        SharedContext ctx = new SharedContext(runId, "ambiguous: 'make links safer' [Java]", RAW_REQUEST);
        Graph graph = buildGraph(Common.PROJECT_ROOT);
        EventLog eventLog = new EventLog(runId, outDir);
        Executor executor = new Executor(graph, ctx, eventLog, new AutoApprover(graph), true);

        RunResult result = executor.run();
        eventLog.close();
        Common.writeReport(result, outDir.resolve("REPORT.md"));
        Common.printSummary(result);

        Map<String, Object> reqOutput = ctx.getOutput("requirements");
        System.out.println("\n--- requirements agent output ---");
        System.out.println("  ambiguity_detected: " + reqOutput.get("ambiguity_detected"));
        System.out.println("  chosen_interpretation: " + reqOutput.get("chosen_interpretation"));
        System.out.println("  rejected_interpretations: " + reqOutput.get("rejected_interpretations"));
        System.out.println("\nFull report: runs/" + runId + "/REPORT.md");
    }
}
